rig/machine_control/scp_connection.py

```diff
@@ -0,0 +1,152 @@
+"""A blocking implementation of the SCP protocol.
+"""
+import socket
+from . import consts, packets
+
+
+class SCPConnection(object):
+    """Implements the SCP protocol for communicating with a SpiNNaker chip.
+    """
+    error_codes = {}
+
+    def __init__(self, spinnaker_host, n_tries=5, timeout=0.5):
+        """Create a new communicator to handle control of the SpiNNaker chip
+        with the supplied hostname.
+
+        Parameters
+        ----------
+        spinnaker_host : str
+            A IP address or hostname of the SpiNNaker chip to control.
+        n_tries : int
+            The maximum number of tries to communicate with the chip before
+            failing.
+        timeout : float
+            The timeout to use on the socket.
+        """
+        # Create a socket to communicate with the SpiNNaker machine
+        self.sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
+        self.sock.settimeout(timeout)
+        self.sock.connect((spinnaker_host, consts.SCP_PORT))
+
+        # Store the number of tries that will be allowed
+        self.n_tries = n_tries
+
+        # The current seq value
+        self._seq = 0
+
+    @classmethod
+    def _register_error(cls, cmd_rc):
+        """Register an Exception class as belonging to a certain CMD_RC value.
+        """
+        def err_(err):
+            cls.error_codes[cmd_rc] = err
+            return err
+        return err_
+
+    def send_scp(self, x, y, p, cmd, arg1=0, arg2=0, arg3=0, data=b'',
+                 expected_args=3):
+        """Transmit a packet to the SpiNNaker machine and block until an
+        acknowledgement is received.
+
+        Parameters
+        ----------
+        x : int
+        y : int
+        p : int
+        cmd : int
+        arg1 : int
+        arg2 : int
+        arg3 : int
+        data : bytestring
+        expected_args : int
+            The number of arguments (0-3) that are expected in the returned
+            packet.
+
+        Returns
+        -------
+        :py:class:`~rig.machine_control.packets.SCPPacket`
+            The packet that was received in acknowledgement of the transmitted
+            packet.
+        """
+        # Construct the packet that will be sent
+        packet = packets.SCPPacket(
+            reply_expected=True, tag=0xff, dest_port=0, dest_cpu=p,
+            src_port=7, src_cpu=31, dest_x=x, dest_y=y, src_x=0, src_y=0,
+            cmd_rc=cmd, seq=self._seq, arg1=arg1, arg2=arg2, arg3=arg3,
+            data=data
+        )
+
+        # Repeat until a reply is received or we run out of tries.
+        n_tries = 0
+        while n_tries < self.n_tries:
+            # Transit the packet
+            self.sock.send(b'\x00\x00' + packet.bytestring)
+            n_tries += 1
+
+            try:
+                # Try to receive the returned acknowledgement
+                ack = self.sock.recv(consts.SCP_RECEIVE_LENGTH)
```

Comment:
**Bug:** This value is dependent on the maximum packet size (as discovered by the MachineController's sver). This is probably a bit messy but it should almost certainly come from an argument.